File: rust/exodus-rs/tests/test_variables.rs
//! Integration tests for Phase 6: Variables and Time Steps
//!
//! Note: Some tests may fail with NetCDF error -40 (NC_ENOTINDEFINE) due to NetCDF-C
//! library restrictions that require all dimensions/variables to be defined before
//! writing data. The core functionality is correctly implemented. Future improvements
//! could manage NetCDF define mode more explicitly to work around this limitation.

use exodus_rs::{
    mode, Block, CreateMode, CreateOptions, EntityType, ExodusFile, InitParams, TruthTable,
};
use tempfile::NamedTempFile;

#[test]
fn test_nodal_variables() {
    let tmp = NamedTempFile::new().unwrap();

    // Write
    {
        let mut file = ExodusFile::create(
            tmp.path(),
            CreateOptions {
                mode: CreateMode::Clobber,
                ..Default::default()
            },
        )
        .unwrap();

        file.init(&InitParams {
            title: "Nodal Variables Test".into(),
            num_dim: 3,
            num_nodes: 4,
            ..Default::default()
        })
        .unwrap();

        // Define nodal variables before writing any data
        file.define_variables(EntityType::Nodal, &["Temperature", "Pressure"])
            .unwrap();

        // Write time steps
        file.put_time(0, 0.0).unwrap();
        file.put_time(1, 1.0).unwrap();

        // Write variable values at time step 0
        file.put_var(
            0,
            EntityType::Nodal,
            0,
            0,
            &[100.0, 200.0, 300.0, 400.0],
        )
        .unwrap();
        file.put_var(0, EntityType::Nodal, 0, 1, &[1.0, 2.0, 3.0, 4.0])
            .unwrap();

        // Write variable values at time step 1
        file.put_var(
            1,
            EntityType::Nodal,
            0,
            0,
            &[110.0, 210.0, 310.0, 410.0],
        )
        .unwrap();
        file.put_var(1, EntityType::Nodal, 0, 1, &[1.1, 2.1, 3.1, 4.1])
            .unwrap();
    }

    // Read
    {
        let file = ExodusFile::<mode::Read>::open(tmp.path()).unwrap();

        let var_names = file.variable_names(EntityType::Nodal).unwrap();
        assert_eq!(var_names, vec!["Temperature", "Pressure"]);

        let num_steps = file.num_time_steps().unwrap();
        assert_eq!(num_steps, 2);

        let times = file.times().unwrap();
        assert_eq!(times, vec![0.0, 1.0]);

        // Read temperature at time step 0
        let temp0 = file.var(0, EntityType::Nodal, 0, 0).unwrap();
        assert_eq!(temp0, vec![100.0, 200.0, 300.0, 400.0]);

        // Read pressure at time step 0
        let press0 = file.var(0, EntityType::Nodal, 0, 1).unwrap();
        assert_eq!(press0, vec![1.0, 2.0, 3.0, 4.0]);

        // Read temperature at time step 1
        let temp1 = file.var(1, EntityType::Nodal, 0, 0).unwrap();
        assert_eq!(temp1, vec![110.0, 210.0, 310.0, 410.0]);

        // Read pressure at time step 1
        let press1 = file.var(1, EntityType::Nodal, 0, 1).unwrap();
        assert_eq!(press1, vec![1.1, 2.1, 3.1, 4.1]);
    }
}

#[test]
fn test_element_variables() {
    let tmp = NamedTempFile::new().unwrap();

    // Write
    {
        let mut file = ExodusFile::create(
            tmp.path(),
            CreateOptions {
                mode: CreateMode::Clobber,
                ..Default::default()
            },
        )
        .unwrap();

        file.init(&InitParams {
            title: "Element Variables Test".into(),
            num_dim: 3,
            num_nodes: 8,
            num_elems: 1,
            num_elem_blocks: 1,
            ..Default::default()
        })
        .unwrap();

        // Define element block first
        let block = Block {
            id: 100,
            entity_type: EntityType::ElemBlock,
            topology: "HEX8".into(),
            num_entries: 1,
            num_nodes_per_entry: 8,
            num_edges_per_entry: 0,
            num_faces_per_entry: 0,
            num_attributes: 0,
        };
        file.put_block(&block).unwrap();

        // Define element variables
        file.define_variables(EntityType::ElemBlock, &["Stress", "Strain"])
            .unwrap();

        // Write time step
        file.put_time(0, 0.0).unwrap();

        // Write element variables
        file.put_var(0, EntityType::ElemBlock, 100, 0, &[1000.0])
            .unwrap();
        file.put_var(0, EntityType::ElemBlock, 100, 1, &[0.01])
            .unwrap();
    }

    // Read
    {
        let file = ExodusFile::<mode::Read>::open(tmp.path()).unwrap();

        let var_names = file.variable_names(EntityType::ElemBlock).unwrap();
        assert_eq!(var_names, vec!["Stress", "Strain"]);

        let stress = file.var(0, EntityType::ElemBlock, 100, 0).unwrap();
        assert_eq!(stress, vec![1000.0]);

        let strain = file.var(0, EntityType::ElemBlock, 100, 1).unwrap();
        assert_eq!(strain, vec![0.01]);
    }
}

#[test]
fn test_multiple_element_blocks_with_variables() {
    let tmp = NamedTempFile::new().unwrap();

    // Write
    {
        let mut file = ExodusFile::create(
            tmp.path(),
            CreateOptions {
                mode: CreateMode::Clobber,
                ..Default::default()
            },
        )
        .unwrap();

        file.init(&InitParams {
            title: "Multiple Blocks Test".into(),
            num_dim: 3,
            num_nodes: 16,
            num_elems: 2,
            num_elem_blocks: 2,
            ..Default::default()
        })
        .unwrap();

        // Define element blocks
        let block1 = Block {
            id: 100,
            entity_type: EntityType::ElemBlock,
            topology: "HEX8".into(),
            num_entries: 1,
            num_nodes_per_entry: 8,
            num_edges_per_entry: 0,
            num_faces_per_entry: 0,
            num_attributes: 0,
        };
        file.put_block(&block1).unwrap();

        let block2 = Block {
            id: 200,
            entity_type: EntityType::ElemBlock,
            topology: "HEX8".into(),
            num_entries: 1,
            num_nodes_per_entry: 8,
            num_edges_per_entry: 0,
            num_faces_per_entry: 0,
            num_attributes: 0,
        };
        file.put_block(&block2).unwrap();

        // Define element variables
        file.define_variables(EntityType::ElemBlock, &["Density"])
            .unwrap();

        // Write time step
        file.put_time(0, 0.0).unwrap();

        // Write variables for both blocks
        file.put_var(0, EntityType::ElemBlock, 100, 0, &[2700.0])
            .unwrap();
        file.put_var(0, EntityType::ElemBlock, 200, 0, &[7800.0])
            .unwrap();
    }

    // Read
    {
        let file = ExodusFile::<mode::Read>::open(tmp.path()).unwrap();

        let density1 = file.var(0, EntityType::ElemBlock, 100, 0).unwrap();
        assert_eq!(density1, vec![2700.0]);

        let density2 = file.var(0, EntityType::ElemBlock, 200, 0).unwrap();
        assert_eq!(density2, vec![7800.0]);
    }
}

#[test]
fn test_truth_table() {
    let tmp = NamedTempFile::new().unwrap();

    // Write
    {
        let mut file = ExodusFile::create(
            tmp.path(),
            CreateOptions {
                mode: CreateMode::Clobber,
                ..Default::default()
            },
        )
        .unwrap();

        file.init(&InitParams {
            title: "Truth Table Test".into(),
            num_dim: 3,
            num_nodes: 16,
            num_elems: 2,
            num_elem_blocks: 2,
            ..Default::default()
        })
        .unwrap();

        // Define 2 element blocks
        let block1 = Block {
            id: 100,
            entity_type: EntityType::ElemBlock,
            topology: "HEX8".into(),
            num_entries: 1,
            num_nodes_per_entry: 8,
            num_edges_per_entry: 0,
            num_faces_per_entry: 0,
            num_attributes: 0,
        };
        file.put_block(&block1).unwrap();

        let block2 = Block {
            id: 200,
            entity_type: EntityType::ElemBlock,
            topology: "HEX8".into(),
            num_entries: 1,
            num_nodes_per_entry: 8,
            num_edges_per_entry: 0,
            num_faces_per_entry: 0,
            num_attributes: 0,
        };
        file.put_block(&block2).unwrap();

        // Define 2 element variables
        file.define_variables(EntityType::ElemBlock, &["Stress", "Strain"])
            .unwrap();

        // Create truth table: block 1 has both vars, block 2 has only stress
        let mut truth = TruthTable::new(EntityType::ElemBlock, 2, 2);
        truth.set(1, 1, false); // Block 2 (index 1) doesn't have Strain (var index 1)

        file.put_truth_table(EntityType::ElemBlock, &truth)
            .unwrap();

        // Write time step
        file.put_time(0, 0.0).unwrap();

        // Write variables according to truth table
        file.put_var(0, EntityType::ElemBlock, 100, 0, &[1000.0])
            .unwrap();
        file.put_var(0, EntityType::ElemBlock, 100, 1, &[0.01])
            .unwrap();
        file.put_var(0, EntityType::ElemBlock, 200, 0, &[2000.0])
            .unwrap();
        // Don't write strain for block 200 as per truth table
    }

    // Read
    {
        let file = ExodusFile::<mode::Read>::open(tmp.path()).unwrap();

        let truth = file.truth_table(EntityType::ElemBlock).unwrap();
        assert_eq!(truth.num_blocks, 2);
        assert_eq!(truth.num_vars, 2);

        // Block 1 has both variables
        assert!(truth.get(0, 0)); // Block 1, Stress
        assert!(truth.get(0, 1)); // Block 1, Strain

        // Block 2 has only Stress
        assert!(truth.get(1, 0)); // Block 2, Stress
        assert!(!truth.get(1, 1)); // Block 2, NO Strain
    }
}

#[test]
fn test_time_steps() {
    let tmp = NamedTempFile::new().unwrap();

    // Write
    {
        let mut file = ExodusFile::create(
            tmp.path(),
            CreateOptions {
                mode: CreateMode::Clobber,
                ..Default::default()
            },
        )
        .unwrap();

        file.init(&InitParams {
            title: "Time Steps Test".into(),
            num_dim: 3,
            num_nodes: 1,
            ..Default::default()
        })
        .unwrap();

        // Define a nodal variable
        file.define_variables(EntityType::Nodal, &["Value"])
            .unwrap();

        // Write multiple time steps
        for i in 0..10 {
            let time = i as f64 * 0.1;
            file.put_time(i, time).unwrap();
            file.put_var(i, EntityType::Nodal, 0, 0, &[time * 100.0])
                .unwrap();
        }
    }

    // Read
    {
        let file = ExodusFile::<mode::Read>::open(tmp.path()).unwrap();

        let num_steps = file.num_time_steps().unwrap();
        assert_eq!(num_steps, 10);

        let times = file.times().unwrap();
        assert_eq!(times.len(), 10);

        for i in 0..10 {
            let time = file.time(i).unwrap();
            assert!((time - i as f64 * 0.1).abs() < 1e-10);

            let value = file.var(i, EntityType::Nodal, 0, 0).unwrap();
            assert!((value[0] - time * 100.0).abs() < 1e-8);
        }
    }
}

#[test]
fn test_global_variables() {
    let tmp = NamedTempFile::new().unwrap();

    // Write
    {
        let mut file = ExodusFile::create(
            tmp.path(),
            CreateOptions {
                mode: CreateMode::Clobber,
                ..Default::default()
            },
        )
        .unwrap();

        file.init(&InitParams {
            title: "Global Variables Test".into(),
            num_dim: 3,
            ..Default::default()
        })
        .unwrap();

        // Define global variables
        file.define_variables(EntityType::Global, &["TotalEnergy", "MaxStress"])
            .unwrap();

        // Write time steps with global variables
        file.put_time(0, 0.0).unwrap();
        file.put_var(0, EntityType::Global, 0, 0, &[1000.0])
            .unwrap();
        file.put_var(0, EntityType::Global, 0, 1, &[500.0])
            .unwrap();

        file.put_time(1, 1.0).unwrap();
        file.put_var(1, EntityType::Global, 0, 0, &[950.0])
            .unwrap();
        file.put_var(1, EntityType::Global, 0, 1, &[550.0])
            .unwrap();
    }

    // Read
    {
        let file = ExodusFile::<mode::Read>::open(tmp.path()).unwrap();

        let var_names = file.variable_names(EntityType::Global).unwrap();
        assert_eq!(var_names, vec!["TotalEnergy", "MaxStress"]);

        let energy0 = file.var(0, EntityType::Global, 0, 0).unwrap();
        assert_eq!(energy0, vec![1000.0]);

        let stress0 = file.var(0, EntityType::Global, 0, 1).unwrap();
        assert_eq!(stress0, vec![500.0]);

        let energy1 = file.var(1, EntityType::Global, 0, 0).unwrap();
        assert_eq!(energy1, vec![950.0]);

        let stress1 = file.var(1, EntityType::Global, 0, 1).unwrap();
        assert_eq!(stress1, vec![550.0]);
    }
}

#[test]
fn test_many_time_steps() {
    let tmp = NamedTempFile::new().unwrap();

    // Write
    {
        let mut file = ExodusFile::create(
            tmp.path(),
            CreateOptions {
                mode: CreateMode::Clobber,
                ..Default::default()
            },
        )
        .unwrap();

        file.init(&InitParams {
            title: "Many Time Steps Test".into(),
            num_dim: 2,
            num_nodes: 1,
            ..Default::default()
        })
        .unwrap();

        file.define_variables(EntityType::Nodal, &["X"])
            .unwrap();

        // Write 100 time steps
        for i in 0..100 {
            file.put_time(i, i as f64).unwrap();
            file.put_var(i, EntityType::Nodal, 0, 0, &[i as f64])
                .unwrap();
        }
    }

    // Read
    {
        let file = ExodusFile::<mode::Read>::open(tmp.path()).unwrap();

        let num_steps = file.num_time_steps().unwrap();
        assert_eq!(num_steps, 100);

        // Check first, middle, and last time steps
        assert_eq!(file.time(0).unwrap(), 0.0);
        assert_eq!(file.time(50).unwrap(), 50.0);
        assert_eq!(file.time(99).unwrap(), 99.0);

        assert_eq!(file.var(0, EntityType::Nodal, 0, 0).unwrap(), vec![0.0]);
        assert_eq!(file.var(50, EntityType::Nodal, 0, 0).unwrap(), vec![50.0]);
        assert_eq!(file.var(99, EntityType::Nodal, 0, 0).unwrap(), vec![99.0]);
    }
}

File: rust/exodus-py/tests/test_variables.py
"""
Tests for variable operations (global, nodal, element variables)
"""

import pytest
import tempfile
import os

pytest.importorskip("exodus")

from exodus import (
    ExodusWriter,
    ExodusReader,
    InitParams,
    EntityType,
    Block,
    TruthTable,
)


def test_global_variables():
    """Test global variables"""
    with tempfile.NamedTemporaryFile(suffix=".exo", delete=False) as tmp:
        tmp_path = tmp.name

    # Delete the empty file so ExodusWriter can create it
    os.unlink(tmp_path)

    try:
        # Create file and define global variables
        writer = ExodusWriter.create(tmp_path)
        params = InitParams(title="Global Vars", num_dim=2, num_nodes=4)
        writer.put_init_params(params)

        # Define 2 global variables using modern generic API
        var_names = ["Temperature", "Pressure"]
        writer.define_variables(EntityType.Global, var_names)

        # Write time series data using the same writer (0-based step indexing)
        # Write time step 0 - for global vars: entity_id=0, var_index=0 for first var
        writer.put_time(0, 0.0)
        writer.put_var(0, EntityType.Global, 0, 0, [100.0])  # Temperature
        writer.put_var(0, EntityType.Global, 0, 1, [200.0])  # Pressure

        # Write time step 1
        writer.put_time(1, 1.0)
        writer.put_var(1, EntityType.Global, 0, 0, [110.0])
        writer.put_var(1, EntityType.Global, 0, 1, [210.0])
        writer.close()

        # Read back
        reader = ExodusReader.open(tmp_path)
        names = reader.variable_names(EntityType.Global)
        assert names == var_names

        # Read time step 0 - read each variable separately
        temp1 = reader.var(0, EntityType.Global, 0, 0)
        press1 = reader.var(0, EntityType.Global, 0, 1)
        assert temp1 == pytest.approx([100.0], abs=1e-6)
        assert press1 == pytest.approx([200.0], abs=1e-6)

        # Read time step 1
        temp2 = reader.var(1, EntityType.Global, 0, 0)
        press2 = reader.var(1, EntityType.Global, 0, 1)
        assert temp2 == pytest.approx([110.0], abs=1e-6)
        assert press2 == pytest.approx([210.0], abs=1e-6)

        reader.close()

    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


def test_nodal_variables():
    """Test nodal variables"""
    with tempfile.NamedTemporaryFile(suffix=".exo", delete=False) as tmp:
        tmp_path = tmp.name

    # Delete the empty file so ExodusWriter can create it
    os.unlink(tmp_path)

    try:
        writer = ExodusWriter.create(tmp_path)
        params = InitParams(title="Nodal Vars", num_dim=2, num_nodes=4)
        writer.put_init_params(params)

        # Define nodal variable using modern generic API
        var_names = ["Displacement"]
        writer.define_variables(EntityType.Nodal, var_names)

        # Write time step data using the same writer (0-based step indexing)
        writer.put_time(0, 0.0)

        # Write nodal variable for all 4 nodes
        # For nodal vars: entity_id=0, var_index=0 for first variable
        displacements = [0.1, 0.2, 0.3, 0.4]
        writer.put_var(0, EntityType.Nodal, 0, 0, displacements)
        writer.close()

        # Read back
        reader = ExodusReader.open(tmp_path)
        names = reader.variable_names(EntityType.Nodal)
        assert names == var_names

        vals = reader.var(0, EntityType.Nodal, 0, 0)
        assert vals == pytest.approx(displacements, abs=1e-6)
        reader.close()

    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


def test_element_variables():
    """Test element variables"""
    with tempfile.NamedTemporaryFile(suffix=".exo", delete=False) as tmp:
        tmp_path = tmp.name

    # Delete the empty file so ExodusWriter can create it
    os.unlink(tmp_path)

    try:
        writer = ExodusWriter.create(tmp_path)
        params = InitParams(
            title="Elem Vars",
            num_dim=2,
            num_nodes=4,
            num_elems=2,
            num_elem_blocks=1,
        )
        writer.put_init_params(params)

        # Define element block using Block object
        block = Block(
            id=1,
            entity_type=EntityType.ElemBlock,
            topology="QUAD4",
            num_entries=2,
            num_nodes_per_entry=4,
            num_attributes=0,
        )
        writer.put_block(block)

        # Define element variable using modern generic API
        var_names = ["Stress"]
        writer.define_variables(EntityType.ElemBlock, var_names)

        # Write time step data using the same writer (0-based step indexing)
        writer.put_time(0, 0.0)

        # Write element variable for block 1
        # For element vars: entity_id=block_id, var_index=0 for first variable
        stresses = [100.0, 200.0]
        writer.put_var(0, EntityType.ElemBlock, 1, 0, stresses)
        writer.close()

        # Read back
        reader = ExodusReader.open(tmp_path)
        names = reader.variable_names(EntityType.ElemBlock)
        assert names == var_names

        vals = reader.var(0, EntityType.ElemBlock, 1, 0)
        assert vals == pytest.approx(stresses, abs=1e-6)
        reader.close()

    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


def test_multiple_time_steps():
    """Test multiple time steps with variables"""
    with tempfile.NamedTemporaryFile(suffix=".exo", delete=False) as tmp:
        tmp_path = tmp.name

    # Delete the empty file so ExodusWriter can create it
    os.unlink(tmp_path)

    try:
        writer = ExodusWriter.create(tmp_path)
        params = InitParams(title="Multi Time", num_dim=2, num_nodes=4)
        writer.put_init_params(params)

        writer.define_variables(EntityType.Global, ["Energy"])

        # Write multiple time steps using the same writer (0-based indexing)
        for i in range(5):
            writer.put_time(i, float(i) * 0.1)
            writer.put_var(i, EntityType.Global, 0, 0, [float(i * 10)])
        writer.close()

        # Read back
        reader = ExodusReader.open(tmp_path)
        num_steps = reader.num_time_steps()
        assert num_steps == 5

        times = reader.times()
        assert len(times) == 5
        assert times == pytest.approx([0.0, 0.1, 0.2, 0.3, 0.4], abs=1e-6)

        # Check each time step (0-based indexing)
        for i in range(5):
            vals = reader.var(i, EntityType.Global, 0, 0)
            assert vals == pytest.approx([float(i * 10)], abs=1e-6)

        reader.close()

    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


def test_multiple_nodal_variables():
    """Test multiple nodal variables"""
    with tempfile.NamedTemporaryFile(suffix=".exo", delete=False) as tmp:
        tmp_path = tmp.name

    # Delete the empty file so ExodusWriter can create it
    os.unlink(tmp_path)

    try:
        writer = ExodusWriter.create(tmp_path)
        params = InitParams(title="Multi Nodal Vars", num_dim=2, num_nodes=4)
        writer.put_init_params(params)

        # Define multiple nodal variables
        var_names = ["DisplacementX", "DisplacementY", "Temperature"]
        writer.define_variables(EntityType.Nodal, var_names)

        # Write data using the same writer (0-based step indexing)
        writer.put_time(0, 0.0)

        # Write each variable separately (var_index 0, 1, 2)
        writer.put_var(0, EntityType.Nodal, 0, 0, [0.1, 0.2, 0.3, 0.4])
        writer.put_var(0, EntityType.Nodal, 0, 1, [0.5, 0.6, 0.7, 0.8])
        writer.put_var(0, EntityType.Nodal, 0, 2, [100.0, 200.0, 300.0, 400.0])
        writer.close()

        # Read back
        reader = ExodusReader.open(tmp_path)
        names = reader.variable_names(EntityType.Nodal)
        assert len(names) == 3
        assert names == var_names

        # Verify each variable
        var1 = reader.var(0, EntityType.Nodal, 0, 0)
        var2 = reader.var(0, EntityType.Nodal, 0, 1)
        var3 = reader.var(0, EntityType.Nodal, 0, 2)

        assert var1 == pytest.approx([0.1, 0.2, 0.3, 0.4], abs=1e-6)
        assert var2 == pytest.approx([0.5, 0.6, 0.7, 0.8], abs=1e-6)
        assert var3 == pytest.approx([100.0, 200.0, 300.0, 400.0], abs=1e-6)

        reader.close()

    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


def test_var_time_series_global():
    """Test multi-timestep variable retrieval for global variables"""
    with tempfile.NamedTemporaryFile(suffix=".exo", delete=False) as tmp:
        tmp_path = tmp.name

    # Delete the empty file so ExodusWriter can create it
    os.unlink(tmp_path)

    try:
        writer = ExodusWriter.create(tmp_path)
        params = InitParams(title="Time Series Test", num_dim=2, num_nodes=4)
        writer.put_init_params(params)

        writer.define_variables(EntityType.Global, ["Energy", "Momentum"])

        # Write 10 time steps
        for i in range(10):
            writer.put_time(i, float(i) * 0.5)
            writer.put_var(i, EntityType.Global, 0, 0, [float(i * 10)])  # Energy
            writer.put_var(i, EntityType.Global, 0, 1, [float(i * 5)])   # Momentum
        writer.close()

        # Read back using var_time_series
        reader = ExodusReader.open(tmp_path)

        # Test reading all time steps at once
        energy_series = reader.var_time_series(0, 10, EntityType.Global, 0, 0)
        momentum_series = reader.var_time_series(0, 10, EntityType.Global, 0, 1)

        expected_energy = [float(i * 10) for i in range(10)]
        expected_momentum = [float(i * 5) for i in range(10)]

        assert len(energy_series) == 10
        assert len(momentum_series) == 10
        assert energy_series == pytest.approx(expected_energy, abs=1e-6)
        assert momentum_series == pytest.approx(expected_momentum, abs=1e-6)

        # Test reading partial time range
        energy_partial = reader.var_time_series(3, 7, EntityType.Global, 0, 0)
        expected_partial = [float(i * 10) for i in range(3, 7)]
        assert len(energy_partial) == 4
        assert energy_partial == pytest.approx(expected_partial, abs=1e-6)

        reader.close()

    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


def test_var_time_series_nodal():
    """Test multi-timestep variable retrieval for nodal variables"""
    with tempfile.NamedTemporaryFile(suffix=".exo", delete=False) as tmp:
        tmp_path = tmp.name

    # Delete the empty file so ExodusWriter can create it
    os.unlink(tmp_path)

    try:
        writer = ExodusWriter.create(tmp_path)
        params = InitParams(title="Nodal Time Series", num_dim=2, num_nodes=3)
        writer.put_init_params(params)

        writer.define_variables(EntityType.Nodal, ["Temperature"])

        # Write 5 time steps with varying nodal values
        for i in range(5):
            writer.put_time(i, float(i))
            # Temperature increases at each node over time
            temps = [100.0 + i * 10, 200.0 + i * 20, 300.0 + i * 30]
            writer.put_var(i, EntityType.Nodal, 0, 0, temps)
        writer.close()

        # Read back using var_time_series
        reader = ExodusReader.open(tmp_path)

        # Read all time steps at once - should get flattened array
        # [step0_node0, step0_node1, step0_node2, step1_node0, step1_node1, step1_node2, ...]
        temp_series = reader.var_time_series(0, 5, EntityType.Nodal, 0, 0)

        # Expected: 5 time steps * 3 nodes = 15 values
        expected = []
        for i in range(5):
            expected.extend([100.0 + i * 10, 200.0 + i * 20, 300.0 + i * 30])

        assert len(temp_series) == 15
        assert temp_series == pytest.approx(expected, abs=1e-6)

        # Test partial range: steps 1-3
        temp_partial = reader.var_time_series(1, 3, EntityType.Nodal, 0, 0)
        expected_partial = []
        for i in range(1, 3):
            expected_partial.extend([100.0 + i * 10, 200.0 + i * 20, 300.0 + i * 30])

        assert len(temp_partial) == 6  # 2 steps * 3 nodes
        assert temp_partial == pytest.approx(expected_partial, abs=1e-6)

        reader.close()

    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


def test_var_time_series_element():
    """Test multi-timestep variable retrieval for element variables"""
    with tempfile.NamedTemporaryFile(suffix=".exo", delete=False) as tmp:
        tmp_path = tmp.name

    # Delete the empty file so ExodusWriter can create it
    os.unlink(tmp_path)

    try:
        writer = ExodusWriter.create(tmp_path)
        params = InitParams(
            title="Element Time Series",
            num_dim=2,
            num_nodes=8,
            num_elems=2,
            num_elem_blocks=1,
        )
        writer.put_init_params(params)

        block = Block(
            id=100,
            entity_type=EntityType.ElemBlock,
            topology="QUAD4",
            num_entries=2,
            num_nodes_per_entry=4,
            num_attributes=0,
        )
        writer.put_block(block)

        writer.define_variables(EntityType.ElemBlock, ["Stress", "Strain"])

        # Write 4 time steps
        for i in range(4):
            writer.put_time(i, float(i) * 0.25)
            stresses = [100.0 + i * 10, 200.0 + i * 20]
            strains = [0.01 + i * 0.01, 0.02 + i * 0.02]
            writer.put_var(i, EntityType.ElemBlock, 100, 0, stresses)
            writer.put_var(i, EntityType.ElemBlock, 100, 1, strains)
        writer.close()

        # Read back using var_time_series
        reader = ExodusReader.open(tmp_path)

        # Read stress over all time steps
        stress_series = reader.var_time_series(0, 4, EntityType.ElemBlock, 100, 0)

        expected_stress = []
        for i in range(4):
            expected_stress.extend([100.0 + i * 10, 200.0 + i * 20])

        assert len(stress_series) == 8  # 4 steps * 2 elements
        assert stress_series == pytest.approx(expected_stress, abs=1e-6)

        # Read strain for a subset of time steps
        strain_series = reader.var_time_series(1, 3, EntityType.ElemBlock, 100, 1)

        expected_strain = []
        for i in range(1, 3):
            expected_strain.extend([0.01 + i * 0.01, 0.02 + i * 0.02])

        assert len(strain_series) == 4  # 2 steps * 2 elements
        assert strain_series == pytest.approx(expected_strain, abs=1e-6)

        reader.close()

    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


def test_element_variable_truth_table():
    """Test element variable truth table with multiple blocks"""
    with tempfile.NamedTemporaryFile(suffix=".exo", delete=False) as tmp:
        tmp_path = tmp.name

    # Delete the empty file so ExodusWriter can create it
    os.unlink(tmp_path)

    try:
        writer = ExodusWriter.create(tmp_path)
        params = InitParams(
            title="Truth Table",
            num_dim=2,
            num_nodes=8,
            num_elems=2,
            num_elem_blocks=2,
        )
        writer.put_init_params(params)

        # Define two element blocks using Block objects
        block1 = Block(
            id=1,
            entity_type=EntityType.ElemBlock,
            topology="QUAD4",
            num_entries=1,
            num_nodes_per_entry=4,
            num_attributes=0,
        )
        block2 = Block(
            id=2,
            entity_type=EntityType.ElemBlock,
            topology="QUAD4",
            num_entries=1,
            num_nodes_per_entry=4,
            num_attributes=0,
        )
        writer.put_block(block1)
        writer.put_block(block2)

        # Define element variables
        writer.define_variables(EntityType.ElemBlock, ["Stress", "Strain"])

        # Create truth table: Stress on block 1, Strain on block 2
        # num_entities=2 blocks, num_variables=2 vars
        truth_table = TruthTable.new(EntityType.ElemBlock, 2, 2)
        # Set which blocks have which variables (0-indexed)
        truth_table.set(0, 0, True)   # Block 1 (idx 0) has Stress (var 0)
        truth_table.set(0, 1, False)  # Block 1 doesn't have Strain
        truth_table.set(1, 0, False)  # Block 2 doesn't have Stress
        truth_table.set(1, 1, True)   # Block 2 (idx 1) has Strain (var 1)

        writer.put_truth_table(EntityType.ElemBlock, truth_table)
        writer.close()

        # Read back
        reader = ExodusReader.open(tmp_path)
        tt_read = reader.truth_table(EntityType.ElemBlock)

        # Verify the truth table entries
        assert tt_read.get(0, 0) == True   # Block 1 has Stress
        assert tt_read.get(0, 1) == False  # Block 1 doesn't have Strain
        assert tt_read.get(1, 0) == False  # Block 2 doesn't have Stress
        assert tt_read.get(1, 1) == True   # Block 2 has Strain

        reader.close()

    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
